### ISSUE TITLE
NSD licence type info does not match with nsd-schema

### ISSUE THREAD

**Issue by srodriguezOPT** [sonata-nfv/son-gkeeper#818]
Currently, the service descriptor exposed by the GK contains a licence_type field that does not match with the licence info showed in the NSD schema.

**Comment by jbonnet** [comment]
@srodriguezOPT 
You are right, but I thought we had agreed that the `licence_type` field would be added, containing one of the values `public`, `owned`, `licenced` or `to buy`? This allows the BSS to build a list of services with different actions possible, while the son-schema, with its `licences: [ {'type':'public'}...]` only allows either `public` or `private`.

Don't you agree? Note: there's an issue with the above implementation I'll add here

**Comment by srodriguezOPT** [comment]
Hi @jbonnet,

It's fine for the BSS but its code is currently aligned with the schemas. The son-schema needs to be changed to allow the "to buy", "owned" and "licensed" options. Do you know who is the responsible for that?

When the schemas are changed we can align the GK and BSS code with them.

**Comment by jbonnet** [comment]
Hi, @srodriguezOPT 
Sorry, I disagree: `son-schema` defines how services and functions are described, while here we're dealing with run-time/instance and more 'commercial' information. A service that is said to be `private` in `son-schema` may have many instances, some `to buy`, one `owned` and others `licensed`. The BSS should follow `son-schema` in what the NSD/VNFD is concerned, but this is not information that `son-schema` needs to describe.

Don't you agree?

**Comment by srodriguezOPT** [comment]
@jbonnet 
ah, ok. I misunderstood you. So, the nsd will contain public/private types and, for the private ones, we will internally consider the owned, licensed and to buy options in the licences.

Anyway, to generate the list of available licences per service, the BSS checks the public/private licence type in the NSD contained in the GK's response (https://github.com/sonata-nfv/son-schema/blob/master/service-descriptor/nsd-schema.yml#L48) but currently the GK responses with no licences field in the NSD section and sends a "licence_type" at root level.

![image](https://user-images.githubusercontent.com/15909508/29513822-bbe60d34-8666-11e7-8ba5-c9ff1d14d91c.png)

Why the GK is not sending the licences field in the NSD? I think that we don't need to "duplicate" this info with another field in the GK's response at root level.

What do you think?

**Comment by jbonnet** [comment]
Hi, @srodriguezOPT 
The GK sends whatever is in the NSD. Since the `licences` field is optional, no currently defined service has it, so there's none.
Please mind that there's no duplication: the `owned`, etc., is a run-time information, and it shouldn't even be mixed with the NSD. I've tried to keep it simple (but maybe I'm wrong! Is this the list where such information should be?), so together with the service descriptor the GK is sending if it is `owned`, ... etc.

**Comment by srodriguezOPT** [comment]
Hi, @jbonnet 

mmm... maybe the licence_type name was confusing for me. Really this field contains the user info related to that service and how to deal with licences:
- public: no licence is required
- owned: no licence is required because the user is the owner of the service
- licenced: licence already acquired by the user
- to buy: a licence request is required for that user

doesn't it?

The BSS was building something similar with the NSD's licences field but this is not needed because the GK is currently doing it with the licence_type.

Ok, I'll adapt it.

**Comment by jbonnet** [comment]
Cool, @srodriguezOPT ! B-)
I can change the name... what's your suggestion?

**Comment by srodriguezOPT** [comment]
what about "user_licence"?

**Comment by jbonnet** [comment]
:+1:

**Comment by jbonnet** [comment]
Closed by #823

**Comment by jbonnet** [comment]
Sorry, @srodriguezOPT , I had closed the issue without first letting you know. Please test from the BSS and let me know if everything is as expected...

**Comment by srodriguezOPT** [comment]
Hi @jbonnet ,

don't worry, I was updating the BSS code. The new code is deployed and seems to work fine with the user_licence field, but we need to test the licence BSS<->GK interaction because currently, the unique NSD deployed is public.

**Comment by srodriguezOPT** [comment]
Hi @jbonnet,

The GK is setting the user_licence field as public even when the licences field contains a private licence:

![image](https://user-images.githubusercontent.com/15909508/29656712-9594174c-88b5-11e7-9de9-358a68fed6d7.png)

**Comment by jbonnet** [comment]
Hi, @srodriguezOPT 
there's an extra `0` under the `licences`... is this the way it should be? We also need an example with more than one licence, to test...

Thanks,

**Comment by srodriguezOPT** [comment]
Hi @jbonnet,

This is only a BSS's way to represent the first element of the array. The used NSD is this: https://github.com/sonata-nfv/son-examples/blob/master/service-projects/sonata-vtc-private-service-sp/sources/nsd/sonata-demo.yml#L21

To include more licences: 

## Licences
licences:
  - type: "private"
    url: "https://sp.int3.sonata-nfv.eu/api/v2/licences"
  - type: "private"
    url: "..."
  - type...

But we talked about to use only one licence, right?

**Comment by jbonnet** [comment]
Hi, @srodriguezOPT 
Yes, we had agreed that only one licence (the 'first' in the array) would be processed, but the code is prepared to accept many licences, I just wanted to test it. But let's make it work with one licence only first.

**Comment by jbonnet** [comment]
@srodriguezOPT 
Can you please send me the (link) package with the licence set, for testing?

**Comment by srodriguezOPT** [comment]
Hi @jbonnet ,

the private package is currently uploaded to the sp.int3 environment. You can see the schema in https://github.com/sonata-nfv/son-examples/blob/master/service-projects/sonata-vtc-private-service-sp/sources/nsd/sonata-demo.yml

**Comment by jbonnet** [comment]
@srodriguezOPT 
I know the schema, but I needed the package... :-D

**Comment by srodriguezOPT** [comment]
@jbonnet 

you can download all packages generated in the son-examples. In this case:
https://jenkins.sonata-nfv.eu/view/DEVELOPMENT/job/son-examples/ws/service-projects/sonata-vtc-private-service-sp.son

**Comment by jbonnet** [comment]
Ok, @srodriguezOPT 
I've made a few tweaks (PR #842), can you please check again?
I'm not happy with the current design, though: we're mixing 'services' with 'service instances', which isn't right.

**Comment by srodriguezOPT** [comment]
Hi @jbonnet 

Same result:

![image](https://user-images.githubusercontent.com/15909508/29791291-b18600fc-8c3c-11e7-904b-715253bba4aa.png)

**Comment by jbonnet** [comment]
Hi, @srodriguezOPT 
On which environment are you getting those values from, `INT3`? 'Cause I can not find NSD with UUID `23224da8...cde2` in the logs...
Thanks

**Comment by srodriguezOPT** [comment]
Hi, @jbonnet 

Yes, INT3. But currently, this is only an NSD deployed, the public one. 

Every time that the int-9-environment-deploy is executed, the uuids change.

**Comment by jbonnet** [comment]
Hi, @srodriguezOPT 
I think PR #852 fixed the bug. Can you please re-test and let me know?
Thanks,

**Comment by srodriguezOPT** [comment]
Hi @jbonnet:

GET https://sp.int3.sonata-nfv.eu/api/v2/services?status=active&limit=10&offset=0

responses with 

```
{
	"error": {
		"code": 500,
		"message": "#<NameError: undefined local variable or method `licensed_element' for #GtkApi:0x0055d0a54b34f0>"
	}
}
```

**Comment by jbonnet** [comment]
Yes... :-1:
Just PR'ing the fix right now, sorry...

**Comment by jbonnet** [comment]
@srodriguezOPT , again please?

**Comment by srodriguezOPT** [comment]
@jbonnet 

now there are no "user_licence" field:

![image](https://user-images.githubusercontent.com/15909508/29874432-08f47660-8d97-11e7-90f1-5312614e72fe.png)

**Comment by jbonnet** [comment]
We're so near, @srodriguezOPT, we're so near!
More logging added... another try, please...?

**Comment by srodriguezOPT** [comment]
@jbonnet 

done. check the logs :)

**Comment by jbonnet** [comment]
It's working now, but I'm deleting the NSD when the user has to buy the service: if the user has the NSD, he/she has everything, why buying a licence?
Is this behaviour right for the BSS?

**Comment by srodriguezOPT** [comment]
@jbonnet 

This option doesn't work. Remember that the BSS uses the info retrieved from the gatekeeper. If the NSD section is deleted, we have no information about the service to instantiate it:

![image](https://user-images.githubusercontent.com/15909508/29878209-2c9fa462-8da2-11e7-95b0-1e3a76ea3d3a.png)

**Comment by jbonnet** [comment]
Ok...
But what parts of the NSD the BSS needs? I see `name`, `description`, `vendor`, `service id` (`licence type` is the `user licence`that the GK injects). What else is needed? Any other field from the NSD for internal processing?

**Comment by srodriguezOPT** [comment]
Yes, these ones, but when the NSD is deleted none of these fields are contained in the GET /services response:

```
       {
		"created_at": "2017-08-30T14:26:21.046+00:00",
		"md5": "8f734a052d8dbb405e0fe0badd314e33",
		"signature": null,
		"status": "active",
		"updated_at": "2017-08-30T14:26:21.046+00:00",
		"username": "sonata-1504103174",
		"uuid": "fd9bec9b-6a90-4ca8-8d68-79181aa7c46b",
		"user_licence": "to buy"
	}
```

And the BSS has no info about the name, description, version... How can the user know what service is licencing for? and how can the user select for instantiation a service if he doesn't know the name and description?

**Comment by jbonnet** [comment]
I'll keep the needed fields within the `nsd` object, but delete the other fields, is that ok?
So the BSS will be able, without any change, to show what it has been showing (except in the details window). 
Agreed?

**Comment by srodriguezOPT** [comment]
yes, this should work.

We only lose the NSD details.

After this, we can try the get /licences api

**Comment by jbonnet** [comment]
@srodriguezOPT 
Done!

**Comment by jbonnet** [comment]
Closed by PR #858
